File: fast-and-slow-pointers/happy-number.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
  public:
    bool isHappy(int n) {
      if (n == 1)
        return true;

      int slow = n, fast = squaredDigitsSum(n);

      while (fast != slow) {
        if (fast == 1)
          return true;

        slow = squaredDigitsSum(slow);
        fast = squaredDigitsSum(
          squaredDigitsSum(fast)
        );
      }

      return false;
    }

  private:
    int squaredDigitsSum(int n) {
      int sum = 0;

      while(n > 0) {
        int dig = n % 10;
        sum += (dig * dig);
        n /= 10;
      }

      return sum;
    }
};

File: fast-and-slow-pointers/happy_number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def digit_square_sum(n: int) -> int:
            sum = 0

            while n:
                rem = n % 10
                sum += rem ** 2
                n //= 10

            return sum

        
        slow = fast = n

        while True:
            slow = digit_square_sum(slow)
            fast = digit_square_sum(digit_square_sum(fast))

            if slow == fast:
                return 1 if slow == 1 else 0